### PR TITLE
Unify pipeline stages fwd and bwd, add pipeline stage tests

### PR DESCRIPTION
**Pull request by H-Huang**
Fixes https://github.com/pytorch/PiPPy/issues/1005

Changes:

- Unify the `forward_one_chunk` and `backward_one_chunk` to be used by both `PipelineStage` and `ManualPipelineStage`
- Add `test_pipeline_stage` to test both `PipelineStage` and `ManualPipelineStage` can be used interchangeably and without error.

TODO:
- create test verify that the outputs for both manual pipeline stage and pipeline stage are equivalent

### CODE REVIEW

File: pippy/ManualPipelineStage.py [hunk 1]
```diff
@@ -359,20 +418,22 @@ def init_p2p_neighbors(self):
     def get_fwd_recv_ops(self) -> List[dist.P2POp]:
         if self.is_first_stage:
             return []
+        self.inputs_grad = [torch.empty_like(inp) for inp in self.inputs]
```
Comment (kwen2501) [inline pippy/ManualPipelineStage.py:323]: Is there a way to avoid creating new tensors at each iteration?

File: pippy/ManualPipelineStage.py [hunk 2]
```diff
@@ -359,20 +418,22 @@ def init_p2p_neighbors(self):
...
         ]
 
     def get_fwd_send_ops(self) -> List[dist.P2POp]:
-        assert (
-            len(self.outputs) != 0
-        ), "forward() must be called before get_fwd_send_ops"
+        output_tuples, flatten_input_tensors = self.fwd_cache[
+            self.fwd_chunk_id - 1
+        ]
+        self.outputs_grad = [torch.empty_like(out) for out in output_tuples]
```
Comment (kwen2501) [inline pippy/ManualPipelineStage.py:333]: Same here.

**Review by kwen2501 — APPROVED** [review]
LGTM! Thanks for the consolidation!
A couple minor comments.

File: pippy/ManualPipelineStage.py [hunk 3]
```diff
@@ -324,6 +330,59 @@ def stage_global_rank(peer_rank):
...
+        """
+        Splits a full-batch input into chunks (i.e. microbatches) and returns
+        the chunks
+        """
+        if args or kwargs:
+            # TODO: cannot split on another dimension other than 0
+            args_split, kwargs_split = split_args_kwargs_into_chunks(
+                args,
+                kwargs,
+                self.num_stages,
```
[the remark this answers is not in the document]
Comment (H-Huang) [reply pippy/ManualPipelineStage.py:269]: good catch, thanks!

**Pull request merged**